contrib/src/main/java/com/datatorrent/contrib/kafka/KafkaTopicOffsetManager.java

```diff
@@ -0,0 +1,487 @@
+   */
+  public void setRetry(int retry)
+  {
+    this.retry = retry;
+  }
+
+  /**
+   * @return the waitTimeout
+   */
+  public long getWaitTimeout()
+  {
+    return waitTimeout;
+  }
+
+  /**
+   * @param waitTimeout
+   *          the waitTimeout to set
+   */
+  public void setWaitTimeout(long waitTimeout)
+  {
+    this.waitTimeout = waitTimeout;
+  }
+
+  public void fetchOffsets()
+  {
+    // version 1 and above fetch from Kafka, version 0 fetches from
+    // ZooKeeper
+
+    OffsetFetchRequest fetchRequest = new OffsetFetchRequest(getConsumerGroup(), getPartitionList(), (short)1,
+        getCorrelationId(), getClientId());
+    try {
+      getChannel().send(fetchRequest.underlying());
+      OffsetFetchResponse fetchResponse = OffsetFetchResponse.readFrom(getChannel().receive().buffer());
+
+      ListIterator<TopicAndPartition> itr = getPartitionList().listIterator();
+
+      //TODO - check and handle more error codes
+
+      while (itr.hasNext()) {
+
+        TopicAndPartition topicAndPartition = itr.next();
+        OffsetMetadataAndError result = fetchResponse.offsets().get(topicAndPartition);
+        short offsetFetchErrorCode = result.error();
+        if (offsetFetchErrorCode == ErrorMapping.NotCoordinatorForConsumerCode()) {
+          getChannel().disconnect();
+
+        } else if (offsetFetchErrorCode == ErrorMapping.BrokerNotAvailableCode()) {
+
+          for (int i = 0; i < getRetryCount(); i++) {
+            // TODO- externalize this wait
+            getChannel().wait(getWaitTimeout());
+            result = fetchResponse.offsets().get(topicAndPartition);
+            // Exit if there is no error
+            if (result.error() == ErrorMapping.NoError()) {
+              break;
+            }
+          }
+          // retry the offset fetch (after backoff)
+        } else {
+          long retrievedOffset = result.offset();
+          // TODO- use the metadata associated with an offset, its not required for offset maintenance as such
+          String retrievedMetadata = result.metadata();
+          getOffsetsMap().put(topicAndPartition, retrievedOffset);
+        }
+
+      }
+
+    } catch (InterruptedException e) {
+      LOG.error("Could not fetch offsets,Exception occurred" + e.getMessage() + "Closing channel");
+      channel.disconnect();
+    }
+
+  }
+
+  @Override
+  public void updateOffsets(Map<KafkaPartition, Long> offsetsOfPartitions)
+  {
+
+    long now = System.currentTimeMillis();
+
+    Map<TopicAndPartition, OffsetAndMetadata> offsets = new LinkedHashMap<TopicAndPartition, OffsetAndMetadata>();
+
+    // Iterate over the updated offsets from the platform's partitions and
+    // create the
+    for (Entry<KafkaPartition, Long> entry : offsetsOfPartitions.entrySet()) {
+      KafkaPartition key = entry.getKey();
+      long value = entry.getValue();
+      offsets.put(new TopicAndPartition(key.getTopic(), key.getPartitionId()), new OffsetAndMetadata(value, "MD", now));
```

Comment:
use entry.getValue() directly instead of declaring local `value` variable as it is used only at one place